homeassistant/components/luftdaten/__init__.py

```diff
@@ -0,0 +1,200 @@
+DEFAULT_ATTRIBUTION = "Data provided by luftdaten.info"
+
+NOTIFICATION_ID = 'luftdaten_notification'
+NOTIFICATION_TITLE = 'Luftdaten Component Setup'
+
+SENSOR_HUMIDITY = 'humidity'
+SENSOR_PM10 = 'P1'
+SENSOR_PM2_5 = 'P2'
+SENSOR_PRESSURE = 'pressure'
+SENSOR_TEMPERATURE = 'temperature'
+
+TOPIC_UPDATE = '{0}_data_update'.format(DOMAIN)
+
+VOLUME_MICROGRAMS_PER_CUBIC_METER = 'µg/m3'
+
+SENSORS = {
+    SENSOR_TEMPERATURE: ['Temperature', 'mdi:thermometer', TEMP_CELSIUS],
+    SENSOR_HUMIDITY: ['Humidity', 'mdi:water-percent', '%'],
+    SENSOR_PRESSURE: ['Pressure', 'mdi:arrow-down-bold', 'Pa'],
+    SENSOR_PM10: ['PM10', 'mdi:thought-bubble',
+                  VOLUME_MICROGRAMS_PER_CUBIC_METER],
+    SENSOR_PM2_5: ['PM2.5', 'mdi:thought-bubble-outline',
+                   VOLUME_MICROGRAMS_PER_CUBIC_METER]
+}
+
+SENSOR_SCHEMA = vol.Schema({
+    vol.Optional(CONF_MONITORED_CONDITIONS, default=list(SENSORS)):
+        vol.All(cv.ensure_list, [vol.In(SENSORS)])
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN:
+        vol.Schema({
+            vol.Required(CONF_SENSOR_ID): cv.positive_int,
+            vol.Optional(CONF_SENSORS, default={}): SENSOR_SCHEMA,
+            vol.Optional(CONF_SHOW_ON_MAP, default=False): cv.boolean,
+            vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL):
+                cv.time_period,
+        })
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Set up the Luftdaten component."""
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][DATA_LUFTDATEN_CLIENT] = {}
+    hass.data[DOMAIN][DATA_LUFTDATEN_LISTENER] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+    station_id = conf.get(CONF_SENSOR_ID)
+
+    if station_id not in configured_sensors(hass):
+        hass.async_add_job(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={'source': SOURCE_IMPORT},
+                data={
+                    CONF_SENSORS: conf[CONF_SENSORS],
+                    CONF_SENSOR_ID: conf[CONF_SENSOR_ID],
+                    CONF_SHOW_ON_MAP: conf[CONF_SHOW_ON_MAP],
+                }
+            )
+        )
+
+    hass.data[DOMAIN][CONF_SCAN_INTERVAL] = conf[CONF_SCAN_INTERVAL]
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up Luftdaten as config entry."""
+    from luftdaten import Luftdaten
+    from luftdaten.exceptions import LuftdatenError
+
+    session = async_get_clientsession(hass)
+
+    try:
+        luftdaten = LuftDatenData(
+            Luftdaten(
+                config_entry.data[CONF_SENSOR_ID], hass.loop, session),
+            config_entry.data.get(CONF_SENSORS, {}).get(
+                    CONF_MONITORED_CONDITIONS, list(SENSORS)))
+        await luftdaten.async_update()
+        hass.data[DOMAIN][DATA_LUFTDATEN_CLIENT][config_entry.entry_id] = \
+            luftdaten
+    except LuftdatenError as err:
```

Comment:
Should we raise `ConfigEntryNotReady` instead ?

Reply:
That would be the next step for config entries as opposed to needing individual implementation of a retry mechanism

Reply:
Added.